enhancements/node-tuning/mixed-cpu-node-plugin.md

```diff
@@ -0,0 +1,316 @@
+`OPENSHIFT_SHARED_CPUS=<CPU-IDs>` will be present under the container's environment variables.
+
+The mixed-cpu-node-plugin itself is essentially a pod that runs as a DaemonSet on the worker nodes.
+cluster-node-tuning-operator (NTO) will be modified to support the deployment of the DaemonSet and additional
+resources (ServiceAccount, RBAC resources, SecurityContextConstraint, etc.)
+NTO will also be responsible to watch, monitor and report the node-plugin's DaemonSet status.
+
+In addition, since both the components are related, mixed-cpu-node-plugin code would be vendor under NTO.
+By having different main.go + container entrypoint, NTO image can be also used as the
+mixed-cpu-node-plugin image. 
+This way saves the overhead of having a separate container image with all the maintenance cost that comes with it.
+
+In order to activate NRI in container-runtime, NTO creates MachineConfig to enable NRI in 
+the container-runtime configuration.
+
+The plugin is completely optional and is off by default.
+
+No changes are needed in the container-runtime, because NRI hooks already supported by the runtime
+and mixed-cpu-node-plugin implementing the NRI interface.
+This minimizes risks and helps avoid regression on
+a container-runtime component (CRI-O in OpenShift case).
+
+### API
+
+At first, the feature is enabled via annotation in the PerformanceProfile:
+
+```yaml
+annotations:
+    mixed-cpu.openshift.io/shared-cpus: "2-3"
+```
+
+The CPU set value should follow the 
+[cpuset](https://github.com/openshift/cluster-node-tuning-operator/blob/master/docs/performanceprofile/performance_profile.md#cpuset) conventions.
+It defines a set of CPUs that can be allocated to guaranteed pods/containers that still require
+non-exclusive ones.
+
+The CPU set must not overlap with `spec.cpu.reserved` or `spec.cpu.isolated` and NTO
+should return an error in case it does.
+
+The reason for using an annotation and not new API field 
+is because there is a coming feature named shared-partition
+that uses similar convention i.e., shared cpus, and would be 
+enabled via NTO as well.
+
+Once shared-partition lands, the annotation should be deprecated
+and node-plugin will use the shared cpus specified 
+at `spec.cpu.shared`.
+
+In case both annotation and `spec.cpu.shared` are specified, 
+the value at `spec.cpu.shared` has precedence over the annotation.
+
+PerformanceProfile example:
+```yaml
+apiVersion: performance.openshift.io/v2
+kind: PerformanceProfile
+metadata:
+  name: example-performance-profile
+  annotations:
+    mixed-cpu.openshift.io/shared-cpus: "2-3"
+spec:
+  cpu:
+    reserved: "0-1"
+    isolated: "4-8"
+  hugepages:
+    defaultHugepagesSize: "1G"
+    pages:
+      - size: "1G"
+        count: 2
+        node: 0
+  realTimeKernel:
+    enabled: true
+  nodeSelector:
+    node-role.kubernetes.io/performance: "test"
+```
+
+Specify the annotation activates the feature, signals NTO to update the `reservedSystemCpus` in Kubelet config, 
```

Comment:
> Can the feature be activated/deactivated on an already running system?

Yes, i'll mention it.

> Would that just require a reboot?

Yes, i'll mention it.

> What happens to existing application pods using CPUs that are moved to the reservedSystemCpus - would they automatically have their affinity updated to move them off those CPUs?

Yes, but it's not related to this feature, hence I didn't specify it.
For example we can achieve the same scenario described above just by changing the `reserved` CPUs under the performance profile.

Reply:
I guess the difference here is that if you deactivate the feature and some pods are already have requests for the new openshift.io/shared-cpus resource, then those pods are going to fail - right? Is that OK? Do we need to prevent that from happening?

Reply:
The pods will keep running until something happen (Kubelet/node restart or pod failures) but then it would go under the normal pod lifecycle process

Reply:
OK - but deactivating this feature is going to require a restart. So that means any pods using the new resource will not come up after the restart - right? And is that OK? I guess we need a big warning in the customer docs?